packages/messaging/src/index.ts

```diff
@@ -527,30 +527,37 @@ namespace MessageLoop {
    *
    * Exceptions in the handler will be caught and logged.
    */
-  function invokeHandler(handler: IMessageHandler, msg: Message): void {
+  function invokeHandler(handler: IMessageHandler, msg: Message): Promise<void> {
     try {
-      handler.processMessage(msg);
+      const retVal = handler.processMessage(msg);
+      if (!retVal) {
+        return Promise.resolve();
+      }
+      return retVal;
     } catch (err) {
       exceptionHandler(err);
     }
+    return Promise.reject("invokeHandler Exception");
   }
 
   /**
    * Add a message to the end of the message queue.
    *
    * This will automatically schedule a run of the message loop.
    */
-  function enqueueMessage(handler: IMessageHandler, msg: Message): void {
-    // Add the posted message to the queue.
-    messageQueue.addLast({ handler, msg });
+  function enqueueMessage(handler: IMessageHandler, msg: Message): Promise<any> {
+    return new Promise((resolve, reject) => {
+      // Add the posted message to the queue.
+      messageQueue.addLast({ handler, msg, resolve, reject });
```

Comment:
Probably the most interesting line in the PR - exposing the resolve and reject callbacks rather than forcing the Promise structure into a paradigm that does not suite it.